$Id: drac_auth.patch,v 1.13 2002/02/19 20:18:02 ken3 Exp $

Patch to add support for Dynamic Relay Authorization Control

For more information about DRAC, see:
	http://mail.cc.umanitoba.ca/drac/index.html


Installation
------------

1.  Apply this patch in the toplevel directory using the following command:

	# patch -b -p0 < contrib/drac_auth.patch

2.  Cleanup any previous builds:

	# make distclean

3a. If you DO NOT have 'smake' and 'autoconf' installed on your system, goto
    step 3b.

    Perform the following to reconfigure your build:

	# rm aclocal.m4 configure config.h.in
	# sh SMakefile
	# ./configure ... --with-drac=<location of libdrac>

    NOTE: you can find your original configure command in config.status

    Proceed to step 4.

3b. Edit imap/Makefile and modify the following three variables:

	DEFS = ... -DDRAC_AUTH
	LIBS = ... -ldrac
	LDFLAGS = ... -L<location of libdrac>

4.  Build and install the software:

	# make
	# make install

5.  If dracd is not running on the same system as Cyrus (localhost),
    use the 'drachost' option in imapd.conf(5) to specify the hostname of
    the dracd server.

6.  Installation is complete!


Operation
---------

The behavior of DRAC is controlled by the value of the 'dracinterval' option
in imapd.conf(5).  If 'dracinterval' is 0 (zero), DRAC support is disabled.
Otherwise, DRAC support is enabled and has the following behavior:

pop3d: Whenever a client opens a user's INBOX, drac_auth() is called.

imapd: Once a client is logged in (via LOGIN or AUTHENTICATE),
       drac_send() will be called once every 'dracinterval' minutes.





Index: acconfig.h
===================================================================
RCS file: /afs/andrew/system/cvs/src/cyrus/acconfig.h,v
retrieving revision 1.28
diff -c -r1.28 acconfig.h
*** acconfig.h	2002/02/18 20:04:16	1.28
--- acconfig.h	2002/02/19 20:06:23
***************
*** 96,101 ****
--- 96,104 ----
  /* do we have SASL support for APOP? */
  #undef HAVE_APOP
  
+ /* the Dynamic Relay Authorization Control package */
+ #undef DRAC_AUTH
+ 
  /* do we have OpenSSL? */
  #undef HAVE_SSL
  
Index: configure.in
===================================================================
RCS file: /afs/andrew/system/cvs/src/cyrus/configure.in,v
retrieving revision 1.219
diff -c -r1.219 configure.in
*** configure.in	2002/02/19 19:52:46	1.219
--- configure.in	2002/02/19 20:06:24
***************
*** 811,816 ****
--- 811,831 ----
  SNMP_SUBDIRS=""
  AC_SUBST(SNMP_SUBDIRS)
  
+ 
+ 
+ dnl
+ dnl Test for DRAC
+ dnl
+ AC_ARG_WITH(drac, [  --with-drac=DIR         use DRAC library in <DIR> [no] ],
+ 	if test -d "$withval"; then
+ 		LDFLAGS="$LDFLAGS -L${withval}"
+ 		AC_CHECK_LIB(drac, dracauth,
+ 			AC_DEFINE(DRAC_AUTH)
+ 			LIBS="${LIBS} -ldrac")
+ 	fi)
+ 
+ 
+ 
  CMU_SOCKETS
  CMU_LIBWRAP
  CMU_UCDSNMP
Index: imap/imapd.c
===================================================================
RCS file: /afs/andrew/system/cvs/src/cyrus/imap/imapd.c,v
retrieving revision 1.342
diff -c -r1.342 imapd.c
*** imap/imapd.c	2002/02/19 18:50:11	1.342
--- imap/imapd.c	2002/02/19 20:06:27
***************
*** 115,120 ****
--- 115,132 ----
  static SSL *tls_conn = NULL;
  #endif /* HAVE_SSL */
  
+ #ifdef DRAC_AUTH
+ static struct {
+     int interval;		/* dracd "ping" interval; 0 = disabled */
+     unsigned long clientaddr;
+     struct prot_waitevent *event;
+ } drac;
+ 
+ extern int dracconn(char *server, char **errmsg);
+ extern int dracsend(unsigned long userip, char **errmsg);
+ extern int dracdisc(char **errmsg);
+ #endif /* DRAC_AUTH */
+ 
  /* current sub-user state */
  static struct mailbox mboxstruct;
  static struct mailbox *imapd_mailbox;
***************
*** 454,459 ****
--- 466,489 ----
      /* setup for sending IMAP IDLE notifications */
      idle_enabled();
  
+ #ifdef DRAC_AUTH
+     /* setup for sending DRAC "pings" */
+     drac.event = NULL;
+     drac.interval = config_getint("dracinterval", 5);
+     if (drac.interval < 0) drac.interval = 0;
+ 
+     if (drac.interval) {
+ 	char *err;
+ 
+ 	if (dracconn((char*) config_getstring("drachost", "localhost"), &err) != 0) {
+ 	    /* disable DRAC */
+ 	    drac.interval = 0;
+ 	    syslog(LOG_NOTICE, "dracconn: %s", err);
+ 	    syslog(LOG_NOTICE, "DRAC DISABLED");
+ 	}
+     }
+ #endif /* DRAC_AUTH */
+ 
      /* create connection to the SNMP listener, if available. */
      snmp_connect(); /* ignore return code */
      snmp_set_str(SERVER_NAME_VERSION,CYRUS_VERSION);
***************
*** 536,541 ****
--- 566,577 ----
  		  imapd_haveaddr = 1;
  	      }
  	}
+ 
+ #ifdef DRAC_AUTH
+ 	drac.clientaddr = imapd_remoteaddr.sin_addr.s_addr;
+     } else {
+ 	drac.clientaddr = 0;
+ #endif /* DRAC_AUTH */
      }
  
      /* create the SASL connection */
***************
*** 578,583 ****
--- 614,624 ----
      prot_flush(imapd_out);
      snmp_increment(ACTIVE_CONNECTIONS, -1);
  
+ #ifdef DRAC_AUTH
+     if (drac.event) prot_removewaitevent(imapd_in, drac.event);
+     drac.event = NULL;
+ #endif /* DRAC_AUTH */
+ 
      /* cleanup */
      imapd_reset();
  
***************
*** 656,661 ****
--- 697,706 ----
  	snmp_increment(ACTIVE_CONNECTIONS, -1);
      }
  
+ #ifdef DRAC_AUTH
+     if (drac.interval) (void) dracdisc((char **)NULL);
+ #endif /* DRAC_AUTH */
+ 
      exit(code);
  }
  
***************
*** 678,684 ****
--- 723,758 ----
  
  }
  
+ #ifdef DRAC_AUTH
  /*
+  * Ping dracd every 'drac.interval' minutes
+  * to let it know that we are still connected
+  */
+ struct prot_waitevent *drac_ping(struct protstream *s,
+ 				 struct prot_waitevent *ev, void *rock)
+ {
+     char *err;
+     static int nfailure = 0;
+ 
+     if (dracsend(drac.clientaddr, &err) != 0) {
+ 	syslog(LOG_NOTICE, "dracsend: %s", err);
+ 	if (++nfailure >= 3) {
+ 	    /* can't contact dracd for 3 consecutive tries - disable DRAC */
+ 	    prot_removewaitevent(s, ev);
+ 	    drac.event = NULL;
+ 	    syslog(LOG_NOTICE, "DRAC DISABLED");
+ 	    return NULL;
+ 	}
+     }
+     else
+ 	nfailure = 0;
+ 
+     ev->mark = time(NULL) + (drac.interval * 60);
+     return ev;
+ }
+ #endif /* DRAC_AUTH */
+ 
+ /*
   * Top-level command loop parsing
   */
  void
***************
*** 1543,1548 ****
--- 1617,1627 ----
  
      prot_printf(imapd_out, "%s OK %s\r\n", tag, reply);
  
+ #ifdef DRAC_AUTH
+     if (drac.interval && drac.clientaddr)
+ 	drac.event = prot_addwaitevent(imapd_in, 0 /* now */, drac_ping, NULL);
+ #endif /* DRAC_AUTH */
+ 
      /* Create telemetry log */
      imapd_logfd = telemetry_log(imapd_userid, imapd_in, imapd_out);
  
***************
*** 1696,1701 ****
--- 1775,1785 ----
  
      prot_setsasl(imapd_in,  imapd_saslconn);
      prot_setsasl(imapd_out, imapd_saslconn);
+ 
+ #ifdef DRAC_AUTH
+     if (drac.interval && drac.clientaddr)
+ 	drac.event = prot_addwaitevent(imapd_in, 0 /* now */, drac_ping, NULL);
+ #endif /* DRAC_AUTH */
  
      /* Create telemetry log */
      imapd_logfd = telemetry_log(imapd_userid, imapd_in, imapd_out);
Index: imap/pop3d.c
===================================================================
RCS file: /afs/andrew/system/cvs/src/cyrus/imap/pop3d.c,v
retrieving revision 1.115
diff -c -r1.115 pop3d.c
*** imap/pop3d.c	2002/02/15 17:21:15	1.115
--- imap/pop3d.c	2002/02/19 20:06:29
***************
*** 101,106 ****
--- 101,110 ----
  
  extern int errno;
  
+ #ifdef DRAC_AUTH
+ static int drac_enabled;
+ extern int dracauth(char *server, unsigned long userip, char **errmsg);
+ #endif /* DRAC_AUTH */
  
  
  #ifdef HAVE_SSL
***************
*** 426,431 ****
--- 430,439 ----
      prot_settimeout(popd_in, timeout*60);
      prot_setflushonread(popd_in, popd_out);
  
+ #ifdef DRAC_AUTH
+     drac_enabled = (config_getint("dracinterval", 5) > 0);
+ #endif /* DRAC_AUTH */
+ 
      if (kflag) kpop();
  
      /* we were connected on pop3s port so we should do 
***************
*** 1385,1390 ****
--- 1393,1411 ----
      }
      popd_mailbox = &mboxstruct;
      proc_register("pop3d", popd_clienthost, popd_userid, popd_mailbox->name);
+ 
+ #ifdef DRAC_AUTH
+     if (drac_enabled) {
+ 	char *err;
+ 
+ 	if (dracauth((char*) config_getstring("drachost", "localhost"),
+ 		     popd_remoteaddr.sin_addr.s_addr, &err) != 0) {
+ 	    /* disable DRAC */
+ 	    drac_enabled = 0;
+ 	    syslog(LOG_NOTICE, "dracauth: %s (DISABLED)", err);
+ 	}
+     }
+ #endif /* DRAC_AUTH */
  
      /* Create telemetry log */
      telemetry_log(popd_userid, popd_in, popd_out);
Index: imap/version.c
===================================================================
RCS file: /afs/andrew/system/cvs/src/cyrus/imap/version.c,v
retrieving revision 1.5
diff -c -r1.5 version.c
*** imap/version.c	2002/02/16 22:00:24	1.5
--- imap/version.c	2002/02/19 20:06:29
***************
*** 131,136 ****
--- 131,140 ----
      snprintf(env_buf + strlen(env_buf), MAXIDVALUELEN - strlen(env_buf),
  	     "; %s", SIEVE_VERSION);
  #endif
+ #ifdef DRAC_AUTH
+     snprintf(env_buf + strlen(env_buf), MAXIDVALUELEN - strlen(env_buf),
+ 	     "; DRAC");
+ #endif
  #ifdef HAVE_LIBWRAP
      snprintf(env_buf + strlen(env_buf), MAXIDVALUELEN - strlen(env_buf),
  	     "; TCP Wrappers");
Index: man/imapd.conf.5
===================================================================
RCS file: /afs/andrew/system/cvs/src/cyrus/man/imapd.conf.5,v
retrieving revision 1.54
diff -c -r1.54 imapd.conf.5
*** man/imapd.conf.5	2002/02/13 17:10:13	1.54
--- man/imapd.conf.5	2002/02/19 20:06:30
***************
*** 308,314 ****
  Unix domain socket that the new mail notification daemon listens on.
  .IP "\fBnotifysocket:\fR /var/imap/socket/notify" 5
  Unix domain socket that the new mail notification daemon listens on.
  .SH SEE ALSO
  .PP
  \fBimapd(8)\fR, \fBpop3d(8)\fR, \fBlmtpd(8)\fR, \fBtimsieved(8)\fR,
! \fBidled(8)\fR, \fBdeliver(8)\fR, \fBmaster(8)\fR
--- 308,320 ----
  Unix domain socket that the new mail notification daemon listens on.
  .IP "\fBnotifysocket:\fR /var/imap/socket/notify" 5
  Unix domain socket that the new mail notification daemon listens on.
+ .IP "\fBdracinterval:\fR 5" 5
+ If nonzero, enables the use of DRAC (Dynamic Relay Authorization Control)
+ by the pop3d and imapd daemons.  Also sets the interval (in minutes) between 
+ re-authorization requests made by imapd.
+ .IP "\fBdrachost:\fR localhost" 5
+ Hostname of the dracd server.
  .SH SEE ALSO
  .PP
  \fBimapd(8)\fR, \fBpop3d(8)\fR, \fBlmtpd(8)\fR, \fBtimsieved(8)\fR,
! \fBidled(8)\fR, \fBdeliver(8)\fR, \fBmaster(8)\fR, \fBrpc.dracd(1m)\fR
